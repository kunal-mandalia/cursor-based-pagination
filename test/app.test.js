/* eslint-disable no-console */
const axios = require('axios').default;

const { start, stop } = require('../app');
const db = require('../context/db');

const port = 8081;
const baseEndpoint = `http://0.0.0.0:${port}`;


function createUsersQuery(first, last, before, after, sort) {
  const limit = first ? `first:${first}` : `last:${last}`;
  const beforeInput = before ? `,before:"${before}"` : '';
  const afterInput = after ? `,after:"${after}"` : '';
  const sortInput = sort ? `,sort: {field:${sort.field},order:${sort.order}}` : '';
  return `
  {
    Users(input: {${limit}${beforeInput}${afterInput}${sortInput}}) {
      edges {
        node {
          name
          age
          created_at
        }
        cursor
      }
      pageInfo {
        endCursor
      }
    }
  }
`;
}

function getNames(users) {
  return users.edges.map((edge) => edge.node.name);
}

async function findUsers({
  first = undefined, last = undefined, before = undefined, after = undefined, sort = undefined,
}, log = false) {
  const query = createUsersQuery(first, last, before, after, sort);
  if (log) { console.log(`query ${query}`); }
  const response = await axios.post(`${baseEndpoint}/graphql`, { query });
  if (response.data.errors) {
    console.error(response.data.errors);
  }
  return response.data.data.Users[0];
}

beforeAll(async () => {
  await start(port);
});

afterAll(async (done) => {
  await stop();
  await db.destroy();
  done();
});

describe('status endpoint', () => {
  it('should report UP', async () => {
    const { data } = await axios.get(`${baseEndpoint}/status`);
    expect(data).toEqual({ status: 'UP' });
  });
});

describe('users query', () => {
  it('should return correct number of users', async () => {
    const users = await findUsers({ first: 3 });
    expect(users.edges.length).toBe(3);
  });

  it('should be ordered by name ascendingly', async () => {
    const input = {
      first: 2,
      sort: { field: 'name', order: 'asc' },
    };
    const users = await findUsers(input);
    expect(users).toMatchSnapshot();
  });

  it('should be ordered by name descendingly', async () => {
    const input = {
      first: 2,
      sort: { field: 'name', order: 'desc' },
    };
    const users = await findUsers(input);
    expect(users).toMatchSnapshot();
  });

  it('should paginate after cursor', async () => {
    const input = {
      first: 2,
      after: '1517443200000___name_Carl_asc',
      sort: { field: 'name', order: 'asc' },
    };
    const users = await findUsers(input);
    expect(getNames(users)).toEqual(['Doug', 'Ed']);
    expect(users).toMatchSnapshot();
  });

  it('should paginate after cursor even with duplicates', async () => {
    const input = {
      first: 2,
      after: '1517443200000___age_22_asc',
      sort: { field: 'age', order: 'asc' },
    };
    const users = await findUsers(input);
    expect(getNames(users)).toEqual(['Ed', 'Doug']);
  });

  it('should paginate after cursor, sorted descendingly', async () => {
    const input = {
      first: 100,
      after: '1517443200000___age_22_desc',
      sort: { field: 'age', order: 'desc' },
    };
    const users = await findUsers(input);
    expect(getNames(users)).toEqual(['Zen', 'Bob', 'Doug', 'Ed']);
  });

  it('should paginate before cursor', async () => {
    const input = {
      first: 2,
      before: '1551398400000',
    };
    const users = await findUsers(input);
    expect(getNames(users)).toEqual(['Carl', 'Ed']);
  });

  it('should return last 2 results', async () => {
    const input = {
      last: 2,
    };
    const users = await findUsers(input);
    expect(getNames(users)).toEqual(['Doug', 'Zen']);
  });

  it('should paginate last 2 results before cursor', async () => {
    const input = {
      last: 2,
      before: '1564617600000',
    };
    const users = await findUsers(input);
    expect(getNames(users)).toEqual(['Alice', 'Bob']);
  });

  it('should paginate last 3 results before cursor with sort', async () => {
    const input = {
      last: 3,
      before: '1583020800000___age_90_asc',
      sort: { field: 'age', order: 'asc' },
    };
    const users = await findUsers(input);
    expect(getNames(users)).toEqual(['Ed', 'Bob', 'Doug']);
  });
});
