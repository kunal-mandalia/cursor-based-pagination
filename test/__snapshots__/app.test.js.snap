// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users query should be ordered by name ascendingly 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "1546300800000___name_Alice_asc",
      "node": Object {
        "age": 20,
        "created_at": "1546300800000",
        "name": "Alice",
      },
    },
    Object {
      "cursor": "1551398400000___name_Bob_asc",
      "node": Object {
        "age": 35,
        "created_at": "1551398400000",
        "name": "Bob",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "1551398400000___name_Bob_asc",
  },
}
`;

exports[`users query should be ordered by name descendingly 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "1583020800000___name_Zen_desc",
      "node": Object {
        "age": 90,
        "created_at": "1583020800000",
        "name": "Zen",
      },
    },
    Object {
      "cursor": "1527811200000___name_Ed_desc",
      "node": Object {
        "age": 22,
        "created_at": "1527811200000",
        "name": "Ed",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "1527811200000___name_Ed_desc",
  },
}
`;

exports[`users query should paginate after cursor 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "1564617600000___name_Doug_asc",
      "node": Object {
        "age": 30,
        "created_at": "1564617600000",
        "name": "Doug",
      },
    },
    Object {
      "cursor": "1527811200000___name_Ed_asc",
      "node": Object {
        "age": 22,
        "created_at": "1527811200000",
        "name": "Ed",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "1527811200000___name_Ed_asc",
  },
}
`;
